### ISSUE TITLE
ARM64 rebuild package not working - always deploys X86

### ISSUE THREAD

**Issue by supertick** [electron-userland/electron-builder#8677]
* **Electron-Builder Version**: 25.1.8
* **Node Version**: 20.18.1
* **Electron Version**: 31.7.3
* **Electron Type (current)**:
* **Target**:  arm64

Trying to cross compile on a github action runner for arm64 (raspi) for serialport bindings
```
  "package": "yarn prebuild && yarn build && yarn copy-public",
  "package-linux-arm64": "yarn package && npx electron-rebuild -f --arch=arm64 && electron-builder --linux AppImage --arm64",

```

The rebuild reports arm64
```
  COPY Release/cpufeatures.node
make: Leaving directory '/home/runner/work/robotlab-x/robotlab-x/server/node_modules/cpu-features/build'
✔ Rebuild Complete
  • electron-builder  version=25.1.8 os=6.5.0-1025-azure
  • loaded configuration  file=package.json ("build" field)
  • @electron/rebuild already used by electron-builder, please consider to remove excess dependency from devDependencies

To ensure your native dependencies are always matched electron version, simply add script `"postinstall": "electron-builder install-app-deps" to your `package.json`
  • executing @electron/rebuild  electronVersion=31.7.3 arch=arm64 buildFromSource=false appDir=./
  • installing native dependencies  arch=arm64
  • preparing       moduleName=@serialport/bindings arch=arm64
  • finished        moduleName=@serialport/bindings arch=arm64
  • preparing       moduleName=@serialport/bindings-cpp arch=arm64
  • finished        moduleName=@serialport/bindings-cpp arch=arm64
  • preparing       moduleName=cpu-features arch=arm64
  • finished        moduleName=cpu-features arch=arm64
  • completed installing native dependencies
  • packaging       platform=linux arch=arm64 electron=31.7.3 appOutDir=/home/runner/work/robotlab-x/robotlab-x/dist/packages/linux-arm64-unpacked
  • downloading     url=https://github.com/electron/electron/releases/download/v31.7.3/electron-v31.7.3-linux-arm64.zip size=106 MB parts=8
  • downloaded      url=https://github.com/electron/electron/releases/download/v31.7.3/electron-v31.7.3-linux-arm64.zip duration=432ms
  • building        target=AppImage arch=arm64 file=/home/runner/work/robotlab-x/robotlab-x/dist/packages/robotlab-x-0.9.224-arm64.AppImage
  • downloading     url=https://github.com/electron-userland/electron-builder-binaries/releases/download/appimage-12.0.1/appimage-12.0.1.7z size=1.6 MB parts=1
  • downloaded      url=https://github.com/electron-userland/electron-builder-binaries/releases/download/appimage-12.0.1/appimage-12.0.1.7z duration=141ms
Done in 25.02s.
```

The AppImage is arm64 but the serialport/bindings are x86:
```
file /tmp/.mount_robotlxD4414/resources/app.asar.unpacked/node_modules/@serialport/bindings/build/Release/bindings.node
/tmp/.mount_robotlxD4414/resources/app.asar.unpacked/node_modules/@serialport/bindings/build/Release/bindings.node: ELF 64-bit LSB shared object, x86-64, version 1 (GNU/Linux), dynamically linked, BuildID[sha1]=b386300dd94cd02c193230bd3b319aee2fe19472, not stripped
```

**Comment by mmaietta** [comment]
Can you try with [v26.0.0-alpha.6](https://github.com/electron-userland/electron-builder/releases/tag/v26.0.0-alpha.6)?

I contributed some changes upstream to `electron/rebuild` and electron-builder updated that dependency in one of the recent alpha versions.

**Comment by supertick** [comment]
Thanks,
I tried v26.0.0-alpha.6, both with npx electron-rebuild -f --arch=arm64 &&  and without
Same behavoir, it appears to correctly build it but the AppImage release contains x86_64 binaries
![image](https://github.com/user-attachments/assets/fe64d3ba-d575-4204-9d61-79f109321015)

**Comment by mmaietta** [comment]
Can you share a minimum reproducible repo that I could take a look at? This sounds like a bug in `@electron/rebuild` that I could investigate and fix upstream.

To hopefully unblock you, you can use the previous app-builder-bin implementation, set `nativeRebuilder: legacy` in your config.

**Comment by supertick** [comment]
The project is currently green shield in github - https://github.com/RobotLab-X/robotlab-x
This is the package.json defining the arm 64 build:
https://github.com/RobotLab-X/robotlab-x/blob/main/server/package.json

Steps to reproduce should be straight forward:
on a linux x86 64 machine:
```bash
git clone https://github.com/RobotLab-X/robotlab-x.git
cd robotlab-x
yarn install-all
yarn build-client
yarn install-client
yarn package-linux-arm64
```
An AppImage will get created in dist/packages.
It will be executable on arm64 - but the file extracted in /tmp/.mount ..../resources/app.asar.unpacked/node_modules /Release/bindings.node will be a x86-64 binary

Thanks for looking into this :)

**Comment by supertick** [comment]
At the moment I have a self-hosted git runner that's building it on a raspi since the cross compiler deployment isn't working

**Comment by bimusiek** [comment]
We have similar issue, the prebuild package is picked incorrectly. It shows as fetching x64 but in fetches arm64.

**Comment by bimusiek** [comment]
@mmaietta I can confirm that alpha.7 version fixed the problem! Thank you!

**Comment by miurahr** [comment]
It works well for AppImage but not for rpm and deb.

```
  • building        target=deb arch=arm64 file=dist/Bluesky-1.0.1-arm64.deb
  • adding autoupdate files for: deb. (Beta feature)  resourceDir=dist/linux-arm64-unpacked/resources
  • downloading     url=https://github.com/electron-userland/electron-builder-binaries/releases/download/fpm-1.9.3-2.3.1-linux-x86/fpm-1.9.3-2.3.1-linux-x86.7z size=4.6 MB parts=1
  • downloaded      url=https://github.com/electron-userland/electron-builder-binaries/releases/download/fpm-1.9.3-2.3.1-linux-x86/fpm-1.9.3-2.3.1-linux-x86.7z duration=3.237s
  ⨯ cannot execute  cause=exit status 255
                    errorOut=i386-binfmt-P: Could not open '/lib/ld-linux.so.2': No such file or directory
```

Please check also for fpm dependency.

**Comment by DaanVervacke** [comment]
> It works well for AppImage but not for rpm and deb.
> 
> ```
>   • building        target=deb arch=arm64 file=dist/Bluesky-1.0.1-arm64.deb
>   • adding autoupdate files for: deb. (Beta feature)  resourceDir=dist/linux-arm64-unpacked/resources
>   • downloading     url=https://github.com/electron-userland/electron-builder-binaries/releases/download/fpm-1.9.3-2.3.1-linux-x86/fpm-1.9.3-2.3.1-linux-x86.7z size=4.6 MB parts=1
>   • downloaded      url=https://github.com/electron-userland/electron-builder-binaries/releases/download/fpm-1.9.3-2.3.1-linux-x86/fpm-1.9.3-2.3.1-linux-x86.7z duration=3.237s
>   ⨯ cannot execute  cause=exit status 255
>                     errorOut=i386-binfmt-P: Could not open '/lib/ld-linux.so.2': No such file or directory
> ```
> 
> Please check also for fpm dependency.

Encountering the same issue.

**Comment by mmaietta** [comment]
@miurahr @DaanVervacke That's not the same issue as OP, please open another issue and add debug logs with `DEBUG=electron-builder` env var set.

Closing per https://github.com/electron-userland/electron-builder/issues/8677#issuecomment-2557313409

**Comment by jimmy-loyola** [comment]
Maybe issue might be occurring because you're not using the correct compilation toolchain. Instead of the standard **/usr/bin/gcc** and **/usr/bin/g++**, you should be using **/usr/bin/aarch64-linux-gnu-gcc** and **/usr/bin/aarch64-linux-gnu-g++** for your specific architecture. To resolve this, you'll need to **apt install gcc-aarch64-linux-gnu g++-aarch64-linux-gnu** packages, and then set the **CC=/usr/bin/aarch64-linux-gnu-gcc** and **CXX=/usr/bin/aarch64-linux-gnu-g++** environment variables accordingly.

Now you should be able to cross compile:

```bash
cd native-module-path
node-gyp --arch=arm64 configure
node-gyp build
```

**Comment by miurahr** [comment]
> [@miurahr](https://github.com/miurahr) [@DaanVervacke](https://github.com/DaanVervacke) That's not the same issue as OP, please open another issue and add debug logs with `DEBUG=electron-builder` env var set.
> 
> Closing per [#8677 (comment)](https://github.com/electron-userland/electron-builder/issues/8677#issuecomment-2557313409)

Sorry it has already been reported as "Fails on arm64 linux Ubuntu, e-b tries to use x86 executable" in July 2021
https://github.com/electron-userland/electron-builder/issues/6116
